### ISSUE TITLE
CNN Context Options Failing

### ISSUE THREAD

**Issue by himanshusagar** [untrix/im2latex#11]
I see 3 values for param "--build-image-contex" in run.py
If I'm not wrong,
A. value 1 of above param, builds vgg context.
B. value 2 of above param, build new cnn architecture to be trained.

Now,
I am getting 
**1. Following error with A(written above) i.e value of param is 1**

 File "/home/{PATH}/im2latex/src/train_multi_gpu.py", line 241, in main
    model = Im2LatexModel(hyper, train_q, opt=opt, reuse=reuse_vars)
  File "/home/{PATH}/im2latex/src/model/Im2LatexModel.py", line 135, in __init__
    self._a = build_vgg_context(params, self._im)
  File "/home/{PATH}/im2latex/src/model/Im2LatexModel.py", line 58, in build_vgg_context
    assert K.int_shape(a)[1:] == (params.H, params.W, params.D)
AssertionError

2. Following Error with B(written above) i.e value of param is 2.

ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[28,64,128,1088]
	 [[Node: Training/gpu_0/I2L/Optimizer/gradients/Training/gpu_0/I2L/Convnet/Convnet_1/MaxPool_2/MaxPool2D/MaxPool_grad/MaxPoolGrad = MaxPoolGrad[T=DT_FLOAT, data_format="NHWC", ksize=[1, 2, 2, 1], padding="SAME", strides=[1, 2, 2, 1], _device="/job:localhost/replica:0/task:0/device:GPU:0"](Training/gpu_0/I2L/Convnet/Convnet_1/Conv_1/Conv/Tanh, Training/gpu_0/I2L/Convnet/Convnet_1/MaxPool_2/MaxPool2D/MaxPool, Training/gpu_0/I2L/Optimizer/gradients/Training/gpu_0/I2L/Convnet/Convnet_1/Conv_3/Conv/Conv2D_grad/tuple/control_dependency)]]
	 [[Node: Training/SyncTowers/optimizer/Adam/update/_1527 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_7501_Training/SyncTowers/optimizer/Adam/update", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

3. I am getting this warning too:
UserWarning: 
Corpus/Sentence contains 0 counts of 2-gram overlaps.
BLEU scores might be undesirable; use SmoothingFunction().

Please tell me way to solve above 3 problems. More importantly, I want to learn ways to reduce Memory usage in your architecture.

**Comment by himanshusagar** [comment]
@untrix Waiting for your opinion.

**Comment by untrix** [comment]
Yes, setting the param to value 1 will build VGG16 which hogs a lot of memory :) If you want to reproduce the model presented in the paper, then leave --build-image-context to the default value (2). Other values build other models which do not work and are therefore not presented in the paper. Hence, you would go with your option 'B' above.

Also, remember to set the num_gpus param in run.py to the number of GPUs you have. It defaults to 2, because I trained on 2 GPUs.

For reducing the memory footprint without changing the model, you can change the batch-size per tower, i.e. the -b option. In my instructions I set it to 28, but you could reduce it further. The final batch-size will be this value * num_gpus. In my case it was 28*2 = 56. However, bear in mind that reducing batch-size reduces the stochasticity of the optimization algorithm and therefore may change the convergence / learning curves. So, you want the maximum batch-size possible. I would suggest to get at least one GPU to train this thing - ideally more than one.

Lastly, if there are any followup questions/comments on this topic, please post those on the wiki.

Note: If you change batch-size to a value that does not divide 56 (assuming you downloaded my data-set which is divided up into bins of size 56), make sure to also give the -p commandline flag. Doing so will make the code ignore the fact that the bins are not integral multiple of the batch-size which means that some samples will get ignored at each epoch (but different samples in the next epoch because they get shuffled before start of an epoch) - something I don't like but that's the only way when you are training in batches. To avoid this problem you should re-bin your data-set into bins of batch-size (or integral multiple of batch-size).

**Comment by himanshusagar** [comment]
Hi @untrix ,
Thanks for your suggestions and prompt reply.
I looked at project [website](https://pcpartpicker.com/user/Sumeet0/saved/#view=gFbvVn), it says you had 2x11Gbx1080s, I have 2x8GBx1080s. Thus, these suggestions should suffice.
Congratulations on your paper!!

**Comment by untrix** [comment]
Hi Himanshu,
     Thanks.
1. If you haven't already done so, I would recommend reading the [paper](https://arxiv.org/abs/1802.05415) before starting to work with the code. That will make things much easier. Yes, I used 2 1080Tis which have 3584 CUDA cores and 11GB memory. Using two 1080s (2560 CUDA cores and 8GB memory) should be perfectly fine - it'll just take longer owing to fewer CUDA cores and you'll need to run with a smaller batch-size in case you run into OOMs (which will further increase the training time).

2. I trained both I2L-NOPOOL as well as I2L-STRIPS for about 100 epochs (as it says in the paper) and that took about 2.5 days on my setup. Then I selected the snapshot with the highest validation bleu2 score for publishing (again, as stated in the paper) and it turned out that that happend at around 147K (standardized) steps for both models - about 2 days into the run. In other words if you were to repeat the experiments with the same data and same parameters, you could expect to hit peak validation bleu2 score in around 2 days - but it is safer to run it for another half day just to be sure. 
- Note: However bear in mind that the validation cycle is only run periodically and therefore you may miss the peak value I got (i.e. bleu2=89.09% for I2L-NOPOOL and 89% for I2L-STRIPS) or you may get a better peak than I got (in case I missed the one you caught). One way to improve your chances of catching the peak is to run the validation cycle more frequently when you approach 146K steps, by either changing the logic in the TrainingLogic.do_validate function (which right now is based on the step/epoch, training bleu2 score of the last 100 steps, max-score recorded so far and some other logic) or simply by manually hitting Ctrl-C :)
- Note: A standardized step is step after adjusting the total batch-size (total of all GPUs) to 64. i.e. if your total batch-size was 128,  1 step = 2 standardized-steps. That's what will show up on tensorboard event-logs and charts (but everywhere else I use the real step). That way the charts of all your experiments will line up correctly on tensorboard (i.e. the epochs will line-up) even if they used different batch-sizes. I used varying batch sizes ranging from 20 to 80 in order to deal with OOM errors and different number of GPUs.

3. The assertion is not a bug (though you should've gotten a different one!) because you are expected to match the H0,D0 and W0 parameters with the expected shape of the convnet's output (corresponds to H', W' and D' in the paper). Look at hyper_params.py:
`        PD(
            'H0', 'Height of feature-map produced by conv-net. Specific to the dataset image size.',
            integer(1),
            LambdaVal(lambda _, p: 4 if (p.image_size == 'big') else 3)
            # LambdaVal(lambda _, p: 8 if (p.build_image_context == 2) else (4 if p.dataset == 3 else 3))
        ),
        PD(
            'W0', 'Width of feature-map produced by conv-net. Specific to the dataset image size.',
            integer(1),
            LambdaVal(lambda _, p: 34 if (p.image_size == 'big') else 33)
            # LambdaVal(lambda _, p: 68 if (p.build_image_context == 2) else (34 if p.dataset == 3 else 33))
        ),
        PD(
            'D0',
            '(integer): number of features coming out of the conv-net. Depth/channels of the last conv-net layer.'
            'See paper or model description.',
            integer(1),
            512),
`
If you change the convnet (e.g. use vgg16 as you are doing or manually change the convent config inside hyper_params.py), then you'll need to set H0, W0 and D0 to appropriate values.
4. But you should've gotten a different assertion so that part is a bug :) **Note:** If you do anything 'non-standard' (i.e. different from the main README instructions) then expect to run into such bugs because this is a research project, not a software product. If you want to go ahead with building vgg16, then you'll need to fix the code to be more like that in build_convnet. I haven't used build_vgg_context for a long time and therefore it is old code and really there is no point to it anymore since you're going to have to tweak the convnet even if you started with vgg16 as a starting point - and in order to do that you'll need to use the build_convnet function i.e. --build-image-context=2. Therefore, I recommend setting --build-image-context=2 and then setting CONVNET in hyper_params.py to the vgg16 architecture. I have already provided the vgg16 architecture there in comments so you could just uncomment that (hyper_params.py lines 907-944. Again, there may be some small bugs here and there which you'll need to fix because I haven't used that code in a while, but at least you get the full vgg16 config in front of you right away. That's what I started with but the performance was pretty awful - if I remember correctly it didn't even begin to learn. So I had to change it significantly to what you see in lines 885-905).
5. In general the code is full of assertions in order to ensure coding errors are caught. There is a long history as to why this specific assertion came about but I won't go into it since it is not relevant now.

Happy experimenting!
-Sumeet
